### ISSUE TITLE
FR: Automatic bug reporting with `jj rage`. 

### ISSUE THREAD

**Issue by PhilipMetzger** [martinvonz/jj#3251]
**Is your feature request related to a problem? Please describe.**
Add a `jj rage` command which copies an anonymized state of the current repo, zips it and allows the user to write a bug report which then creates a new issue in the corresponding issuetracker.  

**Additional context**
I like the term `rage` for bug reporting. 
I'm pretty sure there exists some crate for it already on crates.io. 
I found this while digging through some older Discord conversations.

**Comment by emesterhazy** [comment]
Is this for reporting jj bugs? Maybe something like this should be nested under `jj debug`. Rage is a cheeky name for the command, but `jj debug report` is probably more practical?

I'm also not sure if we'd want to have the bugs automatically filed, or how that would even work since the user would need a Github account to create an issue. Maybe zipping up some debugging info is sufficient?

**Comment by martinvonz** [comment]
For context, Google currently has `hg bugreport` and Meta has `hg rage` AFAIK (I assume that's where Philip got the name from). `jj debug report` seems reasonable (hg doesn't have support for nested subcommands).

**Comment by thoughtpolice** [comment]
I honestly quite like `jj rage` as a name; I thought it was quite cute in both sapling and buck2. I think being a little cheeky is just fine for reporting bugs. :)

> I'm also not sure if we'd want to have the bugs automatically filed, or how that would even work since the user would need a Github account to create an issue. Maybe zipping up some debugging info is sufficient?

I'm not even sure we want to try zipping up the repository, but generally speaking these tools would talk to a proxy server that would relay the bug to GitHub via OAuth if we wanted anonymized reports; alternatively we could support filing bugs with GitHub's API itself assuming that the user has a proper access token on their machine, normally if they use the `gh` tool to login then it's available in something like `$HOME/.config/gh/`; other tools piggyback on `gh` in this way too, like Sapling.

**Comment by necauqua** [comment]
Just the issue URL params are like 90% [there](https://github.com/martinvonz/jj/issues/new?assignees=&labels=&projects=&template=bug_report.md&title=autogenerated&body=hello%20world%0A%0A%3Cdetails%3E%3Csummary%3Elog.txt%3C/summary%3E%0Awhoah,%20logs%3C/details%3E)
Upload the zip somewhere and link it in the issue body

Although I guess it can get super-long and sadly far from every terminal supports OSI 8 code

**Comment by arxanas** [comment]
I'm opposed to the name "rage" as it's not self-descriptive at all.

---

`git-branchless bug-report` includes this:

- System configuration
- Contents of installed Git hooks (relevant for git-branchless)
- Last N events (default 5)
  - Also shows smartlog rendering after each event.
  - All branch/reference names are redacted to consistent identifiers, except for the main branch, which retains its name.
  - Commit hashes are not redacted, which doesn't seem to have been a problem so far, but you could conceivably do consistent-renaming for those as well.

See https://github.com/arxanas/git-branchless/blob/3faa0744cf0b0c7c2fe2bf20e5f441776f21c6dd/git-branchless/tests/test_bug_report.rs for an example.

On several occasions, people have manually redacted information from their smartlogs to discuss something, and I wish I could have told them ahead of time that there's a built-in function for that 🤣.

**Comment by PhilipMetzger** [comment]
Sorry for the slow response here, what I want is something like `git-branchless bug-report` and I'd like to keep it flavorful with the `rage` name. The issuetracker should be configurable though, as Google may point it to their issuetracker, while OSS keeps Github.

**Comment by khionu** [comment]
I'm also not in favour of `rage`, specifically. I'm fine with some names being less serious, but I agree with arxanas on how descriptive it is. `panic` would be more on point. There should also be a fully-serious alias for it, the non-serious name should be the alias.

**Comment by martinvonz** [comment]
I agree with calling about it `jj bugreport`, but I still think the more important question to answer is this:

> It's also a tricky balance between making the bug report useful and anonymous enough. I'll let someone else propose what we would include.

**Comment by ilyagr** [comment]
I will `jj bugreport` if `jj rage` does not reverse-age me. That is clearly what the command is for.

On a more serious note, I prefer calm and researched bug reports, and I feel like `jj rage` (very mildly, but still) encourages the opposite. I'd be more OK with `jj helpmeiamenraged` from this perspective, but that name has other problems.

If we can come up with a name that makes somebody feel like they were just given a fuzzy teddy bear, that would work for me. Maybe `jj sadpuppy` (and/or a unicode version). I suppose that could print an image of a sad puppy and call `jj bugreport`.

**Comment by necauqua** [comment]
`jj sad`, at least as an alias to something serious like `jj bug-report` sounds hilarious \:)

**Comment by ilyagr** [comment]
We can borrow https://github.com/erkin/ponysay.

![image](https://github.com/martinvonz/jj/assets/4123047/473233e2-f039-4c58-8efd-b629652dbd98)

**Comment by PhilipMetzger** [comment]
After hearing all your suggestions, I'm definitely in favor calling it `bugreport` although we probably could have kept `rage` if we had a clear description of what it does. 

And it shouldn't land in the `debug` namespace, as bug reporting should be a toplevel command.

**Comment by necauqua** [comment]
Don't forget the `sad` alias :upside_down_face:

**Comment by khionu** [comment]
> We can borrow https://github.com/erkin/ponysay.

I shared this with a brony friend of mine, and he said that that picture is from a scene where that character is saying "I just don't know what went wrong", and I'm very displeased with how on point that makes it.
